### ISSUE TITLE
Question: "-" "_" should these be special characters?

### ISSUE THREAD

**Issue by thim81** [mesqueeb/case-anything#19]
Hi @mesqueeb,

Sorry to come back to the "special characters" item, when implementating the v2 of case-anything, the special characters results felt a unlogical when it comes to the "-" & "_" characters, that are kept.

When using for example a PascalCase conversion:
```js
const valueAsString = "$sample-example-reports"

const result1 = pascalCase(valueAsString,{keepSpecialCharacters: false}) // result = SampleExampleReports
const result2 = pascalCase(valueAsString,{keepSpecialCharacters: true}) // result = $Sample-Example-Reports
```

result1 is logical but for result2 I would have expected: `$SampleExampleReports`

My first thought would be that the "-" & "_" would be replace since they are linked to the various casing conversion (kebab-case, snake_case, ...) and the special characters would be all the other characters lik "@", "$", ...

**Comment by mesqueeb** [comment]
@thim81 I think it's still early enough to make a breaking change on this version. I see your point. I want to consider it. 

However, wether or not your desired logic is logical or not, depends from point of view. There is also dot and slashes we need to consider, so should we also strip those ? 

We can consider a new prep for the second param with your desired behaviour. That's actually why I made it an object. 

Maybe we can do something like

```js
pascalCase(string, { keepSpecialCharacters: ['$'] })
```

To only keep that special character.
Now we're using the same prop name though, but I have different logic I think : D

Any ideas for propName or syntax ? Thoughts?

**Comment by thim81** [comment]
@mesqueeb 
The definition of "special characters" is indeed vey much linked to the point of view.

I really like the proposal of the 
```
pascalCase(string, { keepSpecialCharacters: ['$'] })
```
Since it gives the implementator the option to define the special characters that will be not stripped from the result.

So in my situation, what would have been the configuration?
```
const valueAsString = "$sample-example-reports"

const result1 = pascalCase(valueAsString,{keepSpecialCharacters: ['$'] }) // result = $SampleExampleReports
const result2 = pascalCase(valueAsString,{keepSpecialCharacters: ['$', '_','-'] }) // result = $SampleExampleReports
```

My 1st thought would be the definition used in "result1", since I consider the "-", "_", ".", "space", not to be special characters, which is off course my point of view, linked to my current implementation.

**Comment by mesqueeb** [comment]
@thim81 please write ```js for your code blocks for syntax highlighting : )

I am thinking either one of these syntaxes:

1. `{ keepCharacters: ['$'] }` — pass which special characters you want to keep
2. `{ removeCharacters: ['-', '_'] }` — pass which special characters you want to remove

In your case your desired result would be achieved like so:

```js
const str = "$sample-example-reports"

// 1. keep $

pascalCase(str, { keepCharacters: ['$'] })
// === '$SampleExampleReports'

// 2. remove - _

pascalCase(str, { remove: ['-', '_'] })
// === '$SampleExampleReports'
```

In order to keep the library small I wanna for the time being just implement one of the two, until there are requests for the other one as well.

Do you understand my proposal ? Which one would you prefer?

**Comment by thim81** [comment]
They both make plenty of sense and will help to solve some less common cases to handle the casing with special characters.

In my case, I would prefer the  `1. keep` option, since the package already removes all the special characters by default.

**Comment by mesqueeb** [comment]
@thim81 ok! start a stopwatch now! : D

**Comment by thim81** [comment]
🕐  started :-)

**Comment by mesqueeb** [comment]
@thim81 in an attempt to implement this in a "smart" way I manage to keep breaking the tests : D

I am out of time now!! Will continue in a couple of hours. 😅

**Comment by mesqueeb** [comment]
@thim81 Took me about 3 hours 😅  I should really learn to code faster.....
BUT I got it done!!

Released as `2.1.1`!

<sub>--<br />Magnetar was made with love by Luca Ban.<br />You cannot sponsor every project, but next time you do, think of <a href="https://github.com/sponsors/mesqueeb">this one</a> for its prolonged maintenance. 💜</sub>

**Comment by thim81** [comment]
@mesqueeb 🙌  🎆  I'll give the latest release a spin.
